### PR TITLE
refactor(table): make block decoder trailer validation fallible

### LINKED ISSUE #184: refactor(table): make block decoder trailer validation fallible
## Context
Found during review of #182.

## Problem
`Decoder::new` currently asserts `restart_interval > 0` for trailer bytes read from disk. On malformed/corrupt SST this can panic in read path instead of returning a structured corruption error.

## Suggested approach
- Introduce a fallible decoder constructor (`try_new`) that validates trailer fields and propagates errors.
- Wire index/data block iterator construction through fallible path where caller can return `crate::Error` instead of panic.
- Add corruption regression tests that tamper trailer restart interval to zero and assert error (no panic).

## Files involved
- `src/table/block/decoder.rs`
- `src/table/index_block/mod.rs`
- `src/table/data_block/mod.rs`
- related table-reader validation paths

Related PR: #182

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Introduce `Decoder::try_new` that returns `Error::InvalidTrailer` on undersized blocks, zero `restart_interval`, invalid `binary_index_step_size`, or corrupt binary/hash-index layout metadata
- Add `Trailer::try_new` size guard to prevent underflow on truncated blocks; `Trailer::new` delegates to `try_new` (panic instead of UB)
- Add `DataBlock::try_iter` / `IndexBlock::try_iter` fallible wrappers and wire all table-reader validation paths through them
- `FullBlockIndex::new` validates block type and trailer at construction time; the pinned filter index does the same at load time
- `DataBlock::point_read` validates the trailer once via `try_iter` before hash/binary index fast paths (no double validation)
- `Scanner`, `table::Iter`, `TwoLevelBlockIndex::Iter`, and `VolatileBlockIndex::Iter` are poisoned on all error paths so callers cannot silently skip corrupt blocks
- Add corruption regression tests that tamper trailer fields and assert structured error (no panic)

## Test plan

- [x] 1026/1026 tests pass with `cargo nextest run`
- [x] 8 corruption regression tests verify `Error::InvalidTrailer` for tampered blocks
- [x] `cargo clippy --all-features -- -D warnings` passes
- [x] `cargo fmt --all -- --check` passes

## Related

- #193 — make meta block field reads fallible (out of scope for this PR)
- #194 — two-level index scan stops prematurely on empty child partitions (pre-existing, out of scope)
- #195 — blob file metadata corruption regression test (out of scope)
- #196 — make index block bound-cursor helpers fallible (pre-existing API, out of scope)
- #197 — add infallible OwnedIndexBlockIter constructor for pre-validated blocks (out of scope)
- #198 — validate block type on cache-hit path (pre-existing, out of scope)

Closes #184

## Summary by CodeRabbit

* **Bug Fixes**
  * Malformed block trailers now surface explicit errors (instead of silent misses) across reads, point-lookup, index/data iteration, and recovery; operations fail fast and stop after detected corruption.
  * Table recovery and index/filter initialization validate block type and trailer integrity up front.

* **Refactor**
  * Iteration and block/iterator construction paths now return fallible results and enter a terminal error state after an unrecoverable failure, yielding a single error then exhausting.

### CODE REVIEW

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff `scanner.rs:117-135` — exhaust scanner on `fetch_next_block` errors:

Fixed in d6dde989 — `fetch_next_block` error now sets `read_count = block_count` before returning error, same as the `try_new` error path.

File: src/table/iter.rs [hunk 1]
```diff
@@ -89,8 +89,11 @@ impl DoubleEndedIterator for OwnedDataBlockIter {
...
 }
 
-fn create_data_block_reader(block: DataBlock, comparator: SharedComparator) -> OwnedDataBlockIter {
-    OwnedDataBlockIter::new(block, |b| b.iter(comparator))
+fn create_data_block_reader(
+    block: DataBlock,
+    comparator: SharedComparator,
+) -> crate::Result<OwnedDataBlockIter> {
+    OwnedDataBlockIter::try_new(block, |b| b.try_iter(comparator))
 }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/iter.rs]: @coderabbitai Valid finding — however this is a pre-existing pattern in `load_block` (util.rs:57-76) which also returns cached blocks without type validation. This PR didn't introduce the cache-hit fast path. Tracked in #198.

**Comment by polaz** [comment]
Closing to reopen with clean history — PR had 31 review threads across 10+ rounds making it hard to review.
